libvirt/net/net_linux_test.go

```diff
@@ -15,247 +15,476 @@ import (
-	t.Cleanup(func() { iptablesCleanup(t, ipt) })
+		fullReqResp, err := controller.VMStartedBuild(&fullReq)
+		must.NoError(t, err)
+		must.NotNil(t, fullReqResp)
+		must.NotNil(t, fullReqResp.DriverNetwork)
+		must.NotNil(t, fullReqResp.TeardownSpec)
 
-	// Ensure passing a nil request object doesn't cause the function to panic.
-	nilRequestResp, err := mockController.VMStartedBuild(nil)
-	must.ErrorContains(t, err, "no request provided")
-	must.Nil(t, nilRequestResp)
+		must.Eq(t, &drivers.DriverNetwork{IP: "192.168.122.58"}, fullReqResp.DriverNetwork)
 
-	// Ensure passing an empty request object doesn't cause the function to
-	// panic.
-	nilRequestResp, err = mockController.VMStartedBuild(&net.VMStartedBuildRequest{})
-	must.NoError(t, err)
-	must.NotNil(t, nilRequestResp)
-	must.Nil(t, nilRequestResp.TeardownSpec)
-
-	// Pass a request that doesn't contain any configured networks to ensure we
-	// correctly handle that.
-	emptyNetworkRequestResp, err := mockController.VMStartedBuild(&net.VMStartedBuildRequest{
-		NetConfig: &net.NetworkInterfacesConfig{},
-		Resources: &drivers.Resources{},
-	})
-	must.NoError(t, err)
-	must.NotNil(t, emptyNetworkRequestResp)
-	must.Nil(t, emptyNetworkRequestResp.TeardownSpec)
-
-	// Test a correct and full request.
-	fullReq := net.VMStartedBuildRequest{
-		DomainName: "nomad-0ea818bc",
-		Hostname:   "nomad-0ea818bc",
-		Hwaddrs:    []string{"52:54:00:1c:7c:14"},
-		NetConfig: &net.NetworkInterfacesConfig{
-			{
-				Bridge: &net.NetworkInterfaceBridgeConfig{
-					Name:  "virbr0",
-					Ports: []string{"ssh", "nomad"},
-				},
+		expectedTeardownRules := [][]string{
+			{"filter", "NOMAD_VT_FW", "-d", "192.168.122.58", "-p", "tcp",
+				"-m", "state", "--state", "NEW", "-m", "tcp", "--dport",
+				"22", "-j", "ACCEPT",
 			},
-		},
-		Resources: &drivers.Resources{
-			Ports: &nomadstructs.AllocatedPorts{
+			{"nat", "NOMAD_VT_PRT", "-d", "10.0.1.161", "-i", "enp126s0",
+				"-p", "tcp", "-m", "tcp", "--dport", "27494", "-j", "DNAT",
+				"--to-destination", "192.168.122.58:22",
+			},
+			{"filter", "NOMAD_VT_FW", "-d", "192.168.122.58", "-p", "tcp",
+				"-m", "state", "--state", "NEW", "-m", "tcp", "--dport",
+				"4646", "-j", "ACCEPT",
+			},
+			{"nat", "NOMAD_VT_PRT", "-d", "10.0.1.161", "-i", "enp126s0",
+				"-p", "tcp", "-m", "tcp", "--dport", "27512", "-j", "DNAT",
+				"--to-destination", "192.168.122.58:4646",
+			},
+		}
+
+		must.EqFunc(t, expectedTeardownRules, fullReqResp.TeardownSpec.IPTablesRules, func(a, b [][]string) bool {
+			if len(a) != len(b) {
+				return false
+			}
+
+			var found int
+
+			for _, ruleA := range a {
+				for _, ruleB := range b {
+					if !reflect.DeepEqual(ruleA, ruleB) {
+						continue
+					}
+					found++
+				}
+			}
+			return found == len(a)
+		})
+
+		// ensure all expectations in the mocks were called
+		mockConnect.AssertExpectations()
+		mockIptables.AssertExpectations()
+
+	})
+
+	t.Run("direct", func(t *testing.T) {
+		testutil.RequireIPTables(t)
```

Comment:
This pattern of including the mocked and "real but requires root" tests as subtests confused me for a second because the subtests share nothing, but I like it! It means you can run just the main test and at least one version of the test will run.

Comments to that effect would help others who haven't seen this pattern before, but no big deal.